File: symfony4-service/src/Services/Zoo/Service/AnimalsServiceImpl.php
<?php

namespace Services\Zoo\Service;

use Services\Zoo\Exception;
use Services\Zoo\Thrift;

class AnimalsServiceImpl implements AnimalsService
{
    use DataStore;

    public function __construct()
    {
        $this->initDataStore();
    }

    /**
     * Retrieves all animals in the zoo;
     *
     * @return Thrift\AnimalList|null
     */
    public function getAll(): ?Thrift\AnimalList
    {
        return new Thrift\AnimalList(["animals" => $this->getAnimals()]);
    }

    /**
     *
     * @param string $type
     * @param Thrift\Animal|null $animal
     */
    public function addAnimal(string $type, ?Thrift\Animal $animal): void
    {
        $animals = $this->getAnimals();
        $animals[] = $animal;
        $this->saveAnimals($animals);
    }

    /**
     *
     * @param string $type
     *
     * @return Thrift\AnimalList|null
     *
     * @throws Exception\InvalidAnimalTypeException
     */
    public function findAnimalsByType(string $type): ?Thrift\AnimalList
    {
        if ($type === "unsupported") {
            throw new Exception\InvalidAnimalTypeException("Unsupported animal type: $type");
        }

        $animals = $this->getAnimals();
        $filteredAnimals = [];
        foreach ($animals as $animal) {
            if ($animal->type === $type) {
                $filteredAnimals[] = $animal;
            }
        }

        return new Thrift\AnimalList(["animals" => $filteredAnimals]);
    }

    /**
     * Marks all animals as vaccinated
     *
     * @param string $type
     * @param string[]|null $names
     *
     * @throws Exception\AlreadyVaccinatedException
     */
    public function markAsVaccinated(string $type, ?array $names): void
    {
        $animals = $this->getAnimals();
        foreach ($animals as $animal) {
            if ($animal->type === $type && (!$names || in_array($animal->name, $names))) {
                if ($animal->vaccinated) {
                    throw new Exception\AlreadyVaccinatedException(
                        "Animal {$animal->name} has already been vaccinated"
                    );
                }
                $animal->vaccinated = true;
            }
        }
        $this->saveAnimals($animals);
    }

    /**
     * @return Thrift\Animal[]
     */
    private function getAnimals(): array
    {
        return $this->dataStore->get("animals", []);
    }

    /**
     * @param Thrift\Animal[] $animals
     */
    private function saveAnimals(array $animals): void
    {
        $this->dataStore->set("animals", $animals);
    }
}

File: spring-service/src/main/java/com/example/services/zoo/AnimalsServiceImpl.java
package com.example.services.zoo;

import com.example.services.zoo.exception.AlreadyVaccinatedException;
import com.example.services.zoo.exception.InvalidAnimalTypeException;
import com.example.services.zoo.thrift.Animal;
import com.example.services.zoo.thrift.AnimalList;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * This code serves as an example, so the operations aren't thread safe and shouldn't be called simultaneously
 */
@Component
public class AnimalsServiceImpl implements AnimalsService {
    private final AnimalList animals = new AnimalList(new ArrayList<>());

    @Override
    public AnimalList getAll() {
        return animals;
    }

    @Override
    public void addAnimal(String type, Animal animal) {
        animals.addToAnimals(animal);
    }

    @Override
    public AnimalList findAnimalsByType(String type) throws InvalidAnimalTypeException {
        return new AnimalList(
                animals.getAnimals().stream()
                        .filter(animal -> animal.getType().equals(type))
                        .collect(Collectors.toList())
        );
    }

    @Override
    public void markAsVaccinated(String type, Optional<List<String>> names) throws AlreadyVaccinatedException {
        List<String> namesList = names.orElse(Collections.emptyList());
        boolean repeatedVaccinationRequested = animals.getAnimals().stream()
                .filter(animal -> animalMatches(animal, type, namesList))
                .anyMatch(Animal::isVaccinated);
        if (repeatedVaccinationRequested) {
            throw new AlreadyVaccinatedException();
        }

        animals.getAnimals().stream()
                .filter(animal -> animalMatches(animal, type, namesList))
                .forEach(animal -> animal.setVaccinated(true));
    }

    private boolean animalMatches(Animal animal, String type, List<String> possibleNamesList) {
        return animal.getType().equals(type) && possibleNamesList.contains(animal.getName());
    }
}
